reaper.go

```diff
@@ -0,0 +1,191 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/opencontainers/runc/libcontainer"
+	"golang.org/x/sys/unix"
+)
+
+type reaper struct {
+	sync.RWMutex
+
+	chansLock     sync.RWMutex
+	exitCodeChans map[int]chan int
+}
+
+func exitStatus(status unix.WaitStatus) int {
+	if status.Signaled() {
+		return exitSignalOffset + int(status.Signal())
+	}
+
+	return status.ExitStatus()
+}
+
+func (r *reaper) getExitCodeCh(pid int) (chan int, error) {
+	r.chansLock.RLock()
+	defer r.chansLock.RUnlock()
+
+	exitCodeCh, exist := r.exitCodeChans[pid]
+	if !exist {
+		return nil, fmt.Errorf("Process %d not found", pid)
+	}
+
+	return exitCodeCh, nil
+}
+
+func (r *reaper) setExitCodeCh(pid int, exitCodeCh chan int) {
+	r.chansLock.Lock()
+	defer r.chansLock.Unlock()
+
+	r.exitCodeChans[pid] = exitCodeCh
+}
+
+func (r *reaper) deleteExitCodeCh(pid int) {
+	r.chansLock.Lock()
+	defer r.chansLock.Unlock()
+
+	exitCodeCh, exist := r.exitCodeChans[pid]
+	if !exist {
+		return
+	}
+
+	close(exitCodeCh)
+
+	delete(r.exitCodeChans, pid)
+}
+
+func (r *reaper) reap() error {
+	var (
+		ws  unix.WaitStatus
+		rus unix.Rusage
+	)
+
+	// When running new processes, libcontainer expects to wait
+	// for the first process actually spawning the container.
+	// This lock allows any code starting a new process to take
+	// the lock prior to the start of this new process, preventing
+	// the subreaper from reaping unwanted processes.
+	r.Lock()
+	defer r.Unlock()
+
+	for {
+		pid, err := unix.Wait4(-1, &ws, unix.WNOHANG, &rus)
+		if err != nil {
+			if err == unix.ECHILD {
+				return nil
+			}
+
+			return err
+		}
+		if pid < 1 {
+			return nil
+		}
+
+		status := exitStatus(ws)
+
+		agentLog.Infof("SIGCHLD pid %d, status %d", pid, status)
+
+		exitCodeCh, err := r.getExitCodeCh(pid)
+		if err != nil {
```

Comment:
Can you make it clear in the comments  that we are only interested in the exit code of the Container and Exec processes, and ignoring the exit code from children and grandchildren(just reaping them)

Reply:
Yes good point, I'll try to add more comments !